src/ipa_pc/mod.rs

```diff
@@ -1100,15 +1317,26 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
                 hiding_bound = std::cmp::max(hiding_bound, cur_poly.hiding_bound());
                 poly += (*coeff, cur_poly.polynomial());
 
-                combined_rand += &(cur_rand.rand * coeff);
+                let randomness = cur_rand.randomness();
+                let commitment = cur_comm.commitment();
+
+                let mut combined_rand_lc = G::ScalarField::zero();
+                let mut combined_comm_lc = <G::Projective as ProjectiveCurve>::zero();
+
+                for (i, (&rand_single, &comm_single)) in randomness.rand.iter().zip(commitment.comm.iter()).enumerate() {
+                    let is = i * key_len;
+                    combined_rand_lc += rand_single * &coeff.pow(&[is as u64]);
+                    combined_comm_lc += &comm_single.mul(coeff.pow(&[is as u64]));
+                }
```

Comment:
We combine the commitments in the wrong way here.
To open a linear combination LC of (segmented) polynomials we need to combine the segments as follows:
Suppose we want to open the LC = c_1 * p_1(X) + c_2 * p_2(X) + c_3 * p_3(X) and assume that the polynomials have the following different number of segments:

p_1(X)    with comm  (C_1[0],C_1[1], C_1[2])
p_2(X)    with comm  (C_2[0])
p_3(X)    with comm  (C_3[0],C_3[1]).

Then the comm of the LC has also 3 segments (C[0], C[1], C[2]), which are the LC's of the segments
 
C[0] = c_1 * C_1[0] + c_2 * C_2[0] + c_3 * C_3[0],  
C[1] = c_1 * C_1[1] + c_2 * 0         + c_3 *  0,
C[2] = c_1 * C_1[2] + c_2 * C_2[2] + c_3 * 0,  
  
I wonder why the opening proof did work out, as we did treat polynomials and their commitments differently.

Reply:
Beside that, the strategy how degree bounds are treated in the context of LCs seems to be wrong. 
It combines(!) the shifted commitments according to the coefficients in the LC, and these coefficients are no random challenges (they are even forced to be equal to one), and sets this combined commitment as the shifted commitment of the combination. This is only correct if all polynomials in an LC have the same degree bound.

For now, it is secure only if a SINGLE polynomial involved in an LC has a degree bound, as it combines(!) the shifted commitments with coefficients which are no random challenges.
If one wants to address the general case I'd opt to remove the artificial constraint that the coefficient of a polynomial, for which we want to prove a degree bound, is equal to 1. Instead we must sample a random challenge (in the Fiat-Shamir sense) lambda and do the random linear combination of 
LC(p_1(X),...) + lambda * (p_i(X) - query_point^ *shifted_p_i(X))

Reply:
I've removed the "combinations" functions from ipa_pc module, but the default implementations of the functions from lib.rs are not working properly, UTs are failed with it. I've checked the original arkworks repo - same thing, default implementation of these function is not applicable to the ipa_pc. For now UTs which use the combinations are commented